### ISSUE TITLE
Using yolov5(m) to train custom data, resulting in high misidentification

### ISSUE THREAD

**Issue by qybing** [ultralytics/yolov5#8964]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Using yolov5(m) to train custom data, resulting in high misidentification, do you have any good suggestions?

### Additional

no

**Comment by MartinPedersenpp** [comment]
Add more training data perhaps?
I think you need to provide some more information about what you are training to detect and how big your dataset is before anyone can help you.

**Comment by qybing** [comment]
> 👋你好@qybing，感谢您对YOLOv5的关注🚀！请访问我们的⭐️ [入门教程](https://docs.ultralytics.com/yolov5)，您可以在其中找到[自定义数据训练](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)等简单任务的快速入门指南，一直到超参数[进化](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)等高级概念。
> 
> 如果这是一个🐛错误报告，请提供截图和**最小可行代码来重现您的问题**，否则我们无法帮助您。
> 
> 如果这是自定义培训❓问题，请提供尽可能多的信息，包括数据集图像、训练日志、屏幕截图以及在线[W&B 日志记录](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#visualize)的公共链接（如果有）。
> 
> 有关业务查询或专业支持请求，请访问https://ultralytics.com或发送电子邮件至[support@ultralytics.com](mailto:support@ultralytics.com)。
> 
> ## 要求
> [**Python>=3.7.0**](https://www.python.org/)安装了所有[requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt)，包括[**PyTorch>=1.7**](https://pytorch.org/get-started/locally/)。开始：
> 
> ```shell
> git clone https://github.com/ultralytics/yolov5  # clone
> cd yolov5
> pip install -r requirements.txt  # install
> ```
> 
> ## 环境
> YOLOv5 可以在以下任何经过验证的最新环境中运行（预装所有依赖项，包括[CUDA](https://developer.nvidia.com/cuda) / [CUDNN](https://developer.nvidia.com/cudnn)、[Python](https://www.python.org/)和[PyTorch](https://pytorch.org/)）：
> 
> * ****带有免费 GPU 的**Google Colab 和 Kaggle笔记本：**[![在 Colab 中打开](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![在 Kaggle 中打开](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5)
> * **谷歌云**深度学习虚拟机。请参阅[GCP 快速入门指南](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
> * **亚马逊**深度学习 AMI。请参阅[AWS 快速入门指南](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
> * **泊坞窗图像**。请参阅[Docker 快速入门指南](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) [![码头工人拉](https://camo.githubusercontent.com/280faedaf431e4c0c24fdb30ec00a66d627404e5c4c498210d3f014dd58c2c7e/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f796f6c6f76353f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/yolov5)
> 
> ## 地位
> [![CI CPU 测试](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg)](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml)
> 
> 如果此标志为绿色，则所有[YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions)持续集成 (CI) 测试当前均通过。CI 测试每 24 小时在 macOS、Windows 和 Ubuntu 上验证YOLOv5 训练 ( [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) )、验证 ( [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) )、推理 ( [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) ) 和导出 ( [export.py ) 的正确操作，并在每次提交时验证一次。](https://github.com/ultralytics/yolov5/blob/master/export.py)
I sent you an email

**Comment by glenn-jocher** [comment]
@qybing 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!
